### PR TITLE
AR6 Climate assessment on reporting

### PR DESCRIPTION
**Pull request by gabriel-abrahao**
## Purpose of this PR
This PR adds the option to run AR6's [climate-assessment](https://github.com/iiasa/climate-assessment) as part of the normal `output.R` procedure, under the option `ar6Climate`. If one wants it to be run automatically, it can be added to `cfg$output`. This includes the probabilistic assessment of temperature change, compatible with the AR6 standards. 

By default, the whole 600 member MAGICC v7.5.3 ensemble is run, which can take a while. Only certain percentiles are output, with each percentile in a different variable. Although the original output is yearly, which can be found inside a subfolder `climate-temp`, only data for the REMIND timesteps are added to the standard MIF (`REMIND_generic***`). All new variables have the `AR6 climate diagnostics|` prefix.  The original temperature variable, `Temperature|Global Mean`, is still calculated with the old (v6) MAGICC. The comparable variable in the new assessment is  `AR6 climate diagnostics|Surface Temperature (GSAT)|MAGICCv7.5.3|XXth percentile`, with separate variables for 10 different percentiles.

The subfolder `climate-temp` includes the full output of `climate-assessment`, including files describing temperature exceedence probabilities and the consequent climate classification of the scenario (C1, C2, etc). 

## Type of change
- [x] New feature 

## Checklist:

- [x] My code follows the [coding etiquette](https://github.com/remindmodel/remind/blob/develop/main.gms#L80)
- [x] I performed a self-review of my own code
- [x] I explained my changes within the PR, particularly in hard-to-understand areas
- [x] I checked that the [in-code documentation](https://github.com/remindmodel/remind/blob/develop/main.gms#L120) is up-to-date
- [x] I adjusted the reporting in [`remind2`](https://github.com/pik-piam/remind2) where it was needed
- [x] I adjusted `forbiddenColumnNames` in [readCheckScenarioConfig.R](https://github.com/remindmodel/remind/blob/develop/scripts/start/readCheckScenarioConfig.R) in case the PR leads to deprecated switches
- [ ] All automated model tests pass (`FAIL 0` in the output of `make test`)
- [x] The changelog `CHANGELOG.md` [has been updated correctly](https://gitlab.pik-potsdam.de/rse/rsewiki/-/wikis/Standards-for-Writing-a-Changelog)

### CODE REVIEW

**Comment by gabriel-abrahao** [comment]
There are still two big pain points that should probably be addressed before this is merged. Could someone familiar with Python and our `venv` setup take a look at this please? Maybe @pfuehrlich-pik ?

1. The script makes use of [scripts in climate-assessment](https://github.com/iiasa/climate-assessment/tree/main/scripts). However, the `scripts` folder does not seem to be included when installing the package with `pip`. I've opened [an issue at the package](https://github.com/iiasa/climate-assessment/issues/48) regarding this, but maybe there's a simpler solution. Either way, we need to assure that the version of the scripts and the version of the package used is the one installed in the `venv`. `climate-assessment` is already part of `requirements.txt`, but in my tests it seemed to be using the one in my own personal environment. Currenlty the script works, but I've [hardcoded a scripts folder](https://github.com/gabriel-abrahao/remind/blob/051e54761027490e5c50710cd7d647aa18a05248/scripts/output/single/ar6Climate.R#L105) that definitely shouldn't be there.

2. The script is pretty slow as it is, but it can be substantially sped up if more processors are available simply by [changing this line](https://github.com/gabriel-abrahao/remind/blob/051e54761027490e5c50710cd7d647aa18a05248/scripts/output/single/ar6Climate.R#L89). However, how to determine and/or set up this within the current `output.R` setup? Maybe @LaviniaBaumstark  or @fbenke-pik know?

**Comment by pfuehrlich-pik** [comment]
Hi Gabriel, I don't really know much about python venvs, Mika was our expert here. However, it seems like you are using [run_harm_inf.py](https://github.com/iiasa/climate-assessment/blob/main/scripts/run_harm_inf.py) and [run_clim.py](https://github.com/iiasa/climate-assessment/blob/main/scripts/run_clim.py) which are both quite trivial scripts. They `import climate_assessment.cli` and then call a single function from there. This should also be possible from the pip-installed package (or is it called module in python?), because that includes `cli.py` (it seems like [this subfolder](https://github.com/iiasa/climate-assessment/tree/main/src/climate_assessment) is included for pip-installations). So maybe you don't actually need the scripts, but can import cli directly and call the functions yourself?

File: scripts/output/single/ar6Climate.R [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+if(!exists("source_include")) {
+   # Define arguments that can be read from command line
+   outputdir <- "."
+   readArgs("outputdir", "gdx_name", "gdx_ref_name", "gdx_refpolicycost_name")
+}
+
+gdx           <- file.path(outputdir, gdx_name)
+cfg_path      <- file.path(outputdir, cfg_name)
+logfile       <- file.path(outputdir, "climate.log") # specific log for python steps
```
Comment (orichters) [inline scripts/output/single/ar6Climate.R:22]: Maybe I would prefer `log_climate.txt`, similar to `log_renv.txt`, `log_output.txt` etc.

File: scripts/output/single/ar6Climate.R [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+scenario <- getScenNames(outputdir)
+remind_reporting_file <- file.path(outputdir,paste0("REMIND_generic_", scenario,".mif"))
+
+print(getwd())
+############################# PREPARING EMISSIONS INPUT #############################
+
+# Read the cfg to get the location of MAGICC-related files
+cfg <- read_yaml(cfg_path)
```
Comment (orichters) [inline scripts/output/single/ar6Climate.R:30]: that works? :) I think most assessments load the cfg from `config.Rdata`, but I have no personal preference on that.

File: scripts/output/single/ar6Climate.R [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+# Read the GDX and run reportEmi
+# gdxpath <- "fulldata.gdx"
+cat(date()," ar6Climate.R: Running reportEmi \n")
```
Comment (orichters) [inline scripts/output/single/ar6Climate.R:34]: Why do we need to rerun reportEmi, why not load all the data from the mif file? We can then add a `yaml` file which contains all the variables MAGICC needs in the format
```
- Emissions|BC:
    unit: Mt BC/yr
- Emissions|BC|AFOLU:
    unit: Mt BC/yr
```
and pass that file that we put somewhere in piamInterfaces or REMIND repo as `iiasatemplate` to `generateIIASASubmission` and you only get back what you need.

File: scripts/output/single/ar6Climate.R [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+emimag <- reportEmi(gdx)
+
+# Convert to quitte and add metadata
+emimif <- as.quitte(emimag)
+emimif["scenario"] <- scenario #TODO: Get scenario name from cfg
```
Comment (orichters) [inline scripts/output/single/ar6Climate.R:39]: `cfg$title` should do it, right?

File: scripts/output/single/ar6Climate.R [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+# Write the raw emissions mif
+# TODO: This wouldn't be necessary if we added an option to generateIIASASubmission
```
Comment (orichters) [inline scripts/output/single/ar6Climate.R:42]: generateIIASASubmission should work with a quitte object:
https://github.com/pik-piam/piamInterfaces/blob/master/R/generateIIASASubmission.R#L5-L6
```
> #' @param mifs path to mif files or directories with mif files of a REMIND run,
> #'             or quitte object
```
and if you set `outputFilename` to `NULL`, the function also returns a quitte object, so you can continue working with that directly.

So I think
```
ar6mif <- remind_reporting_file %>%
  filter(region %in% c("GLO", "World")) %>%
  generateIIASASubmission(mapping = "AR6", outputFilename = NULL, logFile = NULL, iiasatempate = "magicc_variables.yaml") %>%
  mutate(region = factor("World"))
```
should basically do what you need. Why do you prefer to pass a csv, not an xlsx? Then, you could just use `write.IAMCxlsx` from the `quitte` package to write it.

File: scripts/output/single/ar6Climate.R [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+magicc_bin_file          <- file.path(cfg$climate_assessment_files_dir,"/magicc-v7.5.3/bin/magicc")
+
+# Create working folder for climate-assessment files
+workfolder <- file.path(outputdir, "climate-temp")
+dir.create(workfolder, showWarnings = F)
+
+# Set relevant environment variables and create a MAGICC worker directory
+Sys.setenv(MAGICC_EXECUTABLE_7=magicc_bin_file)
+Sys.setenv(MAGICC_WORKER_ROOT_DIR=normalizePath(paste0(workfolder,"/workers/"))) # Has to be an absolute path
+Sys.setenv(MAGICC_WORKER_NUMBER=1) # TODO: Get this from slurm or nproc
```
Comment (orichters) [inline scripts/output/single/ar6Climate.R:89]: I suggest:
```
tasks <- gsub("[^0-9].*", "", gsub("^.*--tasks-per-node=", "", cfg$slurmConfig))
tasks <- if (grepl("^[0-9]+$", tasks)) tasks else 1
```

**Review by orichters — COMMENTED** [review]
Nice work, Gabriel, thanks…

My first set of comments ;)

**Comment by fbenke-pik** [comment]
Concerning this line
`Sys.setenv(MAGICC_WORKER_NUMBER=1)`

You should be able get this number by acessing "SLURM_NTASKS" on the cluster. A failsafe solution could look like this:
`workerNumber <- as.numeric(Sys.getenv("SLURM_NTASKS", unset = 1))`

**Comment by orichters** [comment]
@gabriel-abrahao: Merge this to improve slurm options in output.R: https://github.com/gabriel-abrahao/remind/pull/1

File: scripts/output/single/ar6Climate.R (hunk 1 above)
Comment (tonnrueter) [reply scripts/output/single/ar6Climate.R:22]: Done

File: output.R [hunk 7]
```diff
@@ -83,17 +83,21 @@ if ("--help" %in% flags) {
...
                      "--qos=priority --mem=8000", "--qos=short --mem=8000",
                      "--qos=standby --mem=8000", "--qos=priority --mem=32000")
+  
+  # Modify slurm options for ar6 reporting, since we want to run MAGICC in parallel and we'll need a lot of memory
+  if ("ar6Climate" %in% output) slurm_options <- paste(slurm_options[1:3], "--tasks-per-node=12 --mem=32000")
 
   if (!isSlurmAvailable())
     return("direct")
 
   if (!is.null(slurmExceptions)) {
```
Comment (orichters) [inline output.R:97]: something went missing here. `slurmExceptions` is never defined, and the `    slurmExceptions <- if ("reporting" %in% output) "--mem=[0-9]*[0-9]{3}" else NULL` part deleted below is not reproduced anymore.

File: output.R [hunk 8]
```diff
@@ -86,11 +86,18 @@ if ("--help" %in% flags) {
 choose_slurmConfig_output <- function(output) {
   slurm_options <- c("--qos=priority", "--qos=short", "--qos=standby",
                      "--qos=priority --mem=8000", "--qos=short --mem=8000",
-                     "--qos=standby --mem=8000", "--qos=priority --mem=32000",
-                     "direct")
-  if ("reporting" %in% output) slurm_options <- unique(c(grep("--mem=[0-9]*[0-9]{3}", slurm_options, value = TRUE), "direct"))
```
Comment (orichters) [inline output.R]: that line must stay, I think

File: output.R [hunk 9]
```diff
@@ -67,7 +67,7 @@ library(lucode2)
 library(gms)
 require(stringr, quietly = TRUE)
```
Comment (orichters) [inline output.R:69]: please undo this change.

**Review by orichters — CHANGES_REQUESTED** [review]

File: output.R (hunk 7 above)
Comment (tonnrueter) [reply output.R:97]: Yea I'm afraid messed up the merge here. Imma fix it

**Review by orichters — CHANGES_REQUESTED** [review]
I noted some minor things that have changed since Gabriel started working on it and that should be kept.

**Review by orichters — COMMENTED** [review]
Because I created a new merge conflict, I have fixed the stuff that I noted.

File: output.R [hunk 10]
```diff
@@ -83,17 +84,19 @@ if ("--help" %in% flags) {
...
                      "--qos=standby --mem=16000", "--qos=priority --mem=32000")
 
   if (!isSlurmAvailable())
     return("direct")
 
-  if (!is.null(slurmExceptions)) {
-    slurm_options <- grep(slurmExceptions, slurm_options, value = TRUE)
-  }
+  # Modify slurm options for ar6 reporting, since we want to run MAGICC in parallel and we'll need a lot of memory
+  if ("ar6Climate" %in% output) slurm_options <- combine_slurmConfig(slurm_options[1:3], "--tasks-per-node=12 --mem=32000")
```
Comment (orichters) [inline output.R:96]:
```suggestion
if ("ar6Climate" %in% output) slurm_options <- paste(slurm_options[1:3], "--tasks-per-node=12 --mem=32000")
```

File: scripts/output/single/ar6Climate.R [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+logmsg <- paste0(
+  date(), "  ar6Climate.R: Wrote REMIND emission data to '", climateAssessmentEmi, "' for climate-assessment\n"
+)
+cat(logmsg)
+capture.output(cat(logmsg), file = logFile, append = TRUE)
+
+############################# PYTHON/MAGICC SETUP #############################
+
+# Read the cfg to get the location of MAGICC-related files
+cfg <- read_yaml(cfgPath)
```
Comment (orichters) [inline scripts/output/single/ar6Climate.R:98]: Ok to add that? People will definitely want to run that on old runs.
```suggestion
cfg <- read_yaml(cfgPath)

if (is.null(cfg$climate_assessment_root)) cfg$climate_assessment_root <- "/p/projects/rd3mod/python/climate-assessment/src/"
if (is.null(cfg$climate_assessment_files_dir)) cfg$climate_assessment_files_dir <- "/p/projects/rd3mod/climate-assessment-files/"
if (is.null(cfg$cfg$climate_assessment_magicc_bin)) cfg$climate_assessment_magicc_bin <- "/p/projects/rd3mod/climate-assessment-files/magicc-v7.5.3/bin/magicc"
```

**Comment by orichters** [comment]
Some results for `/p/tmp/oliverr/remind-smallfix/output/SSP2EU-NPi_2024-02-20_15.40.25`
```
   period MAGICC6 MAGICC7.5.3
    <int>   <dbl>       <dbl>
 1   2005   0.881       0.868
 2   2010   0.895       0.976
 3   2015   0.999       1.12
 4   2020   1.13        1.23
 5   2025   1.27        1.37
 6   2030   1.40        1.48
 7   2035   1.55        1.62
 8   2040   1.68        1.75
 9   2045   1.81        1.86
10   2050   1.93        2.00
11   2055   2.03        2.10
12   2060   2.13        2.22
13   2070   2.30        2.46
14   2080   2.45        2.66
15   2090   2.58        2.83
16   2100   2.68        2.97
17   2110   2.77       NA
18   2130   2.90       NA
19   2150   3.02       NA
```

**Comment by orichters** [comment]
Comparison for NGFS:
```
   scenario         period `MAGICC7.5.3|50th` MAGICC6
   <fct>             <int>              <dbl>   <dbl>
 1 Current Policies   2010              0.976   0.953
 2 Current Policies   2020              1.24    1.16
 3 Current Policies   2030              1.49    1.40
 4 Current Policies   2040              1.76    1.65
 5 Current Policies   2050              2.04    1.92
 6 Current Policies   2060              2.29    2.18
 7 Current Policies   2070              2.54    2.41
 8 Current Policies   2080              2.77    2.63
 9 Current Policies   2090              2.98    2.84
10 Current Policies   2100              3.20    3.04
11 NDCs               2010              0.976   0.953
12 NDCs               2020              1.24    1.16
13 NDCs               2030              1.50    1.41
14 NDCs               2040              1.73    1.64
15 NDCs               2050              1.90    1.83
16 NDCs               2060              2.04    1.99
17 NDCs               2070              2.16    2.12
18 NDCs               2080              2.26    2.23
19 NDCs               2090              2.33    2.32
20 NDCs               2100              2.39    2.40
21 Net Zero 2050      2010              0.976   0.953
22 Net Zero 2050      2020              1.24    1.16
23 Net Zero 2050      2030              1.48    1.42
24 Net Zero 2050      2040              1.56    1.54
25 Net Zero 2050      2050              1.55    1.58
26 Net Zero 2050      2060              1.50    1.58
27 Net Zero 2050      2070              1.47    1.57
28 Net Zero 2050      2080              1.43    1.57
29 Net Zero 2050      2090              1.40    1.56
30 Net Zero 2050      2100              1.38    1.55
```

<details>
<summary>click to expand code!</summary>

```
library(quitte)
library(tidyr)
library(dplyr)

data <- as.quitte("/p/projects/remind/users/oliverr/data/NGFS3.4_IAM_data_R5.xlsx")
vars <- c("Temperature|Global Mean", "AR6 climate diagnostics|Surface Temperature (GSAT)|MAGICCv7.5.3|50.0th Percentile")
scens <- c("Net Zero 2050", "Current Policies", "NDCs")
t <- seq(2010, 2100, 10)
d <- data %>%
  filter(.data$variable %in% vars, .data$scenario %in% scens, .data$model %in% "REMIND NGFS3", .data$period %in% t) %>%
  mutate(unit = "K") %>%
  df_variation() %>%
  mutate(variable = gsub(vars[[1]], "MAGICC6", .data$variable, fixed = TRUE)) %>%
  mutate(variable = gsub(vars[[2]], "MAGICC7.5.3|50th", .data$variable, fixed = TRUE)) %>%
  pivot_wider(names_from = "variable") %>%
  droplevels() %>%
  print(n = 200)
```
</details>

**Comment by orichters** [comment]
I did another test. I took a REMIND NGFS run from 2023, ran the ar6climate reporting script, and compare it to the AR6 assessment I have received from IIASA (as NGFS_MAGICC7.5.3):
```
   period REMIND_MAGICC6 REMIND_MAGICC7.5.3 NGFS_MAGICC7.5.3
 1   2010          0.956              0.976            0.976
 2   2020          1.16               1.23             1.23
 3   2030          1.40               1.46             1.48
 4   2040          1.64               1.71             1.73
 5   2050          1.89               1.93             1.96
 6   2060          2.10               2.13             2.16
 7   2070          2.29               2.34             2.35
 8   2080          2.48               2.52             2.53
 9   2090          2.65               2.71             2.69
10   2100          2.81               2.87             2.84
```
There is some slight variation that I don't understand, but overall looks good.

**Review by orichters — APPROVED** [review]
Code is fine from my side. Still might be worth investigating why the results slightly differ, but this is a big step forward, thank you @tonnrueter and @gabriel-abrahao.

@LaviniaBaumstark: Can you have a look as well, please, mostly whether you consider the last table good enough in terms of results.

**Comment by LaviniaBaumstark** [comment]
I would see no reason for not merging. Differences in REMIND_MAGICC6 vs. REMIND_MAGICC7.5.3 is no suprice. Why REMIND_MAGICC7.5.3 vs. IIASA_MAGICC7.5.3 are not the same is maybe related to some inconcistend parameterisation but nothing we need to clarify before merging. We are clode enough.

**Comment by LaviniaBaumstark** [comment]
but: please run make test

**Comment by orichters** [comment]
> but: please run make test

Please run `make test` but temporarily out-comment [this line](https://github.com/remindmodel/remind/blob/ffdf2e4007229682e2ab13ed39b1a53eca916765/tests/testthat/test_06-output.R#L8). Is pointless without it.

**Comment by gabriel-abrahao** [comment]
I agree with Lavinia here, the differences in Oli's tests don't seem large enough to justify not merging. I've put a lot more work into making sure everything is correct in this version here than in the previous one, so if there's an actual mistake, it's more likely that the differences came from fixing it. That said, a lot of segfaults and overflows usually happen in some individual MAGICC parametrizations. The IIASA people told me they are aware of it but didn't really give me an explanation yet, so maybe it's related to that. At the end of the day, that's the official AR6 pipeline, so if there's a problem, everyone else also has it, and they are aware.

**Comment by tonnrueter** [comment]
Thanks y'all for chiming in! I'll run the tests and merge asap
